src/settings/settingsdialog.cpp

```diff
@@ -379,22 +388,53 @@ void SettingsDialog::saveYandexEngineEmotion(int emotion)
 // To play test text
 void SettingsDialog::detectYandexTextLanguage()
 {
-    if (ui->yandexTestSpeechEdit->text().isEmpty()) {
-        QMessageBox::information(this, tr("Nothing to play"), tr("Playback text is empty"));
-        return;
-    }
-
-    m_yandexTranslator->detectLanguage(ui->yandexTestSpeechEdit->text(), QOnlineTranslator::Yandex);
+    detectTestTextLanguage(m_yandexTranslator, QOnlineTranslator::Yandex);
 }
 
 void SettingsDialog::speakYandexTestText()
 {
-    if (m_yandexTranslator->error() != QOnlineTranslator::NoError) {
-        QMessageBox::critical(this, tr("Unable to detect language"), m_yandexTranslator->errorString());
-        return;
+    speakTestText(m_yandexTranslator, QOnlineTranslator::Yandex);
+}
+
+void SettingsDialog::onGoogleLanguageSelectionChanged(int languageIndex)
+{
+    auto const configuredLang = ui->googleLanguageComboBox->itemData(languageIndex).value<QOnlineTranslator::Language>();
+    auto const originalRegion = ui->googlePlayerButtons->regions(QOnlineTranslator::Google)[configuredLang]; // It will be lost after googleRegionComboBox is changed if not stored here
+
+    ui->googleRegionComboBox->clear();
+
+    ui->googleRegionComboBox->addItem(tr("Default region"), QLocale::AnyCountry);
+    for (auto validRegion : QOnlineTts::validRegions(configuredLang)) {
+        if (validRegion == QLocale::China)
+            ui->googleRegionComboBox->addItem(tr("Mandarin (China)"), QLocale::China); // for now there's only one Chinese dialect supported
+        else
+            ui->googleRegionComboBox->addItem(QLocale::countryToString(validRegion), validRegion);
     }
 
-    ui->yandexPlayerButtons->speak(ui->yandexTestSpeechEdit->text(), m_yandexTranslator->sourceLanguage(), QOnlineTranslator::Yandex);
+    if (originalRegion == QLocale::China)
+        ui->googleRegionComboBox->setCurrentText(tr("Mandarin (China)"));
+    else
+        ui->googleRegionComboBox->setCurrentText(QLocale::countryToString(originalRegion));
```

Comment:
What I don't understand is that why you restoring the original region if language is changed. Could you explain it a little more? Maybe in comment.

Reply:
It's due to the fixed bug mentioned [here](https://github.com/crow-translate/crow-translate/pull/473#issue-1358551208) (top).
Without this four lines, if a user has already stored a region preference for a particular language and selected the language again in the first combo box, the setting will become lost because the second combo box has to be refilled, triggering the slot `saveGoogleEngineRegion`, which at the point of `settingsdialog.cpp:404` (or 406?) would reset the preference to default. This isn't logical.

The explanation above is quite involved, I will try to simply its argument and insert as a crisp comment.

Reply:
Oh, I get it! Could you please rename `originalRegion` into `configuredRegion` for clarity?

Also I would use `findData` and set the selected item using the found data.

Reply:
> configuredRegion for clarity?

Or even `langRegion`. Will be enough to understand the code. Because I assumed that you restoring the previous selection :smile: